openedx/core/djangoapps/user_api/accounts/api.py

```diff
@@ -429,32 +584,189 @@ def _validate_username(username):
         AccountUsernameInvalid
 
     """
+    try:
+        _validate_unicode(username)
+    except UnicodeError as invalid_unicode_err:
+        raise AccountUsernameInvalid(str(invalid_unicode_err))
+    _validate_username_type(username)
+    _validate_username_of_good_length(username)
+
+    try:
+        validate_slug(username)
+    except ValidationError:
+        raise AccountUsernameInvalid(
+            u"Username '{username}' must contain only A-Z, a-z, 0-9, -, or _ characters".format(username=username)
+        )
+
+
+def _validate_username_type(username):
+    """Validate the username instance type.
+
+    :param username: The proposed username (unicode).
+    :return: None
+    :raises: AccountUsernameInvalid
+    """
     if not isinstance(username, basestring):
         raise AccountUsernameInvalid(u"Username must be a string")
 
+
+def _validate_username_of_good_length(username):
+    """Validate that the username is less than or equal to
+    the maximum username length allowed, and greater than or
+    equal to the minimum username length allowed.
+
+    :param username: The proposed username (unicode).
+    :return: None
+    :raises: AccountUsernameInvalid
+    """
+    _validate_username_satisfies_min_length(username)
+    _validate_username_satisfies_max_length(username)
```

Comment:
I fold there -- I had the same thought as you, but the real reason I didn't end up abstracting it this way is because I wanted to keep the actual exception class that's raised dependent on the error. But I can just scratch the various `AccountXInvalid`, `AccountXAlreadyExists` type exceptions and squash them into a more generic one and then solve this particular problem generically too.